README.md
[![Build Status](https://travis-ci.org/schul-cloud/schulcloud-calendar.svg?branch=master)](https://travis-ci.org/schul-cloud/schulcloud-calendar) [![Greenkeeper badge](https://badges.greenkeeper.io/schul-cloud/schulcloud-calendar.svg)](https://greenkeeper.io/)

# schulcloud-calendar

[Calendar API](https://schul-cloud.github.io/schulcloud-calendar/#/default)

[Schul-Cloud API Requirements](https://github.com/schulcloud/schulcloud-calendar/blob/master/wiki/schulcloud-api-requirements.md)

[Research of Caldav Modules for Node](https://github.com/schulcloud/schulcloud-calendar/blob/master/wiki/node-caldav-research.md)

[Future work](https://github.com/schulcloud/schulcloud-calendar/blob/master/wiki/future-work.md)

## Setup
### Installation
PostgreSQL/bin/* folder must be accessable through shell/cmd (check environment variables) for this install guide to work. Check functionality with `psql --version` for example.
1. Clone repository
2. Install & start PostgreSQL, set and remember `postgres` user password during install
3. Create a user `node`, without password: `createuser -U postgres --superuser --login node`
4. Create a database named `schulcloud_calendar`: `createdb -U postgres schulcloud_calendar`
5. Connect to it: `psql -d schulcloud_calendar -U postgres`
   (Prompt will change to `schulcloud_calendar=#`)
6. Grant user with `all` permissions for the db: `grant all privileges on database schulcloud_calendar TO node;`
   (Feedback should be: `GRANT`)
7. Install [uuid-ossp](https://www.postgresql.org/docs/current/static/uuid-ossp.html) ([install info](https://www.postgresql.org/message-id/C5EBF511-835E-4F24-A4E4-6CC0119F48E4%40me.com)) extension: `CREATE EXTENSION "uuid-ossp";`
   Feedback should be: `CREATE EXTENSION`
8. Quit psql-Interpreter with Ctrl+C (or open new command prompt) and run `npm install`
9. Import DB data: view below
10. Run node:
   `npm start` for production environment
   `npm run start_local` for local unix environment
   `npm run start_local_win` for local windows environment

Alternative install instructions here: [Kalender Setup](https://docs.schul-cloud.org/display/SCDOK/Setup#Setup-Kalender)

### Example database data
If you get `fe_sendauth: no password supplied` errors on commands, edit your `[installdir]\data\pg_hba.conf` and set `METHOD` to `trust` in all lines.
1. Create tables with `psql -U node -d schulcloud_calendar -a -f schema.sql`
2. Insert example data using `psql -U node -d schulcloud_calendar -a -f example_data.sql`

### Tests
1. Create a database named `schulcloud_calendar_test` and setup as described above
2. Create tables with `psql -U node -d schulcloud_calendar_test -a -f schema.sql`
3. Run `npm test`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Allowed Types of change: `Added`, `Changed`, `Deprecated`, `Removed`, `Fixed`, `Security`

## [25.3.2] - 2021-02-02

### Fixed

- SC-7845 updated changelog github action
- SC-8176 sunday events are possible now

### Added

- SC-8512 - route to fix event duplications & setting IS_MIGRATION in config.js

## [25.3.1] - 2020-12-18

## Added

- SC-8217 - new configurable options for postgres client

## [25.2.0]

### Added

- OPS-1297 - Added Changelog github action

### Changed

- SC-7561 - Update logger


test/testutils/getDate.js
const getDate = (minOffset, baseDate) => new Date((baseDate || new Date()).getTime() + (minOffset * 1000 * 60)).toISOString();

module.exports = getDate;


test/routes/events.integration.test.js
const expect = require('chai').expect;
const nock = require('nock');
const request = require('supertest');

const app = require('../../src/app');
const { SERVER_SCOPES_URI, SCHULCLOUD_BASE_PATH } = require('../../src/config');
const db = require('../../src/infrastructure/databasePromise');
const {
	dbUtils,
	serverMockData: { createOverlayWithDefaultScopes, addCourseScope },
	convertEventToJsonApi,
	getDate,
} = require('../testutils');

/** tests */
describe('routes/events', () => {
	const userId = '59898b4a26ffc20c510cfcf0';
	let resetDB;
	let clearData;
	let server;
	const resolvedServerScopes = createOverlayWithDefaultScopes();

	const addCourseTestEvent = async ({ scopeId = '59cce16281297026d02cde123', summary = 'event1', repeat_until, last_modified, weekday}) => {
		const data = { scopeId, summary, startDate: getDate(-30), endDate: getDate(30), repeat_freq: "WEEKLY", repeat_until, last_modified, weekday};
		return addTestEvents(data);
	}

	const addTestEvents = async ({ scopeId = '59cce16281297026d02cde123', summary, startDate, endDate, repeat_freq, repeat_until, last_modified, weekday}) => {
		const data = {
			courseId: scopeId,
			scopeId,
			summary,
			startDate,
			endDate,
			last_modified,
			frequency: repeat_freq,
			repeat_until,
			weekday,
		};

		await nock(SCHULCLOUD_BASE_PATH)
			.get(uri => uri.includes(SERVER_SCOPES_URI))
			.reply(200, () => {
				return resolvedServerScopes;
			});

			const result = await request(app)
				.post('/events')
				.send(convertEventToJsonApi(data))
				.set('Authorization', userId);

			return result.body.data;

	}

	beforeEach(async () => {
		await nock(SCHULCLOUD_BASE_PATH)
			.get(uri => uri.includes(SERVER_SCOPES_URI))
			.reply(200, () => {
				return resolvedServerScopes;
			});
	});

	before(async () => {
		({ resetDB, clearData } = dbUtils(db));
		await resetDB();
		server = await app.listen(3001);
	});

	after(async () => {
		await clearData();
		await server.close();
	});

	describe('events', () => {
		describe('POST', () => {
			const scopeId = '59cce16281297026d02cde123';
			const courseName = 'post test';
			before(async () => {
				await addCourseScope(resolvedServerScopes, scopeId, courseName, true);
			});

			after(async () => {
				resolvedServerScopes.data = resolvedServerScopes.data.filter(scope => scope.id !== scopeId);
			}); 

			it('create event', async () => {
				const eventData = convertEventToJsonApi({
					courseId: scopeId,
					scopeId,
					summary: courseName,
				});

			   const result = await request(app)
					.post('/events')
					.send(eventData)
					.set('Authorization', userId);
				
				const { data } = result.body;

				expect(data).to.be.an('array').to.have.lengthOf(1);

				const relationships = data[0].relationships['scope-ids'];

				expect(relationships).to.be.an('array').to.have.lengthOf(1);
				expect(relationships[0]).to.be.equal(scopeId);

				expect(data[0].attributes.summary).to.be.equal(courseName);
			});
		});
		/*
		describe('FIND all', () => {
			it('get all for this user', async () => {
			   const result = await request(app)
					.get('/events')
					.query({
						all: true,
					})
					.set('Authorization', userId);

				expect(result.body.data).to.be.an('array').to.have.lengthOf(5);
			});
		}); */
	
		describe('FIND with scope and timebox', () => {
			const scopeId = '59cce16281297026d02abc123';
			const scopeIdThatIsNotRequested = '59cce16281297026d02xyz999';
			const scopeIdWithoutReadPermissions = '59cce16281297026d02xyz000'
			const courseName = 'find time box test';
			let events;
			before(async () => {
				addCourseScope(resolvedServerScopes, scopeId, courseName, true);
				addCourseScope(resolvedServerScopes, scopeIdThatIsNotRequested, courseName, true);

				events = await Promise.all([
					addTestEvents({ scopeId, startDate: getDate(-30), endDate: getDate(30), summary: 'touched start'}),
					addTestEvents({ scopeId, startDate: getDate(15), endDate: getDate(45), summary: 'in time'}),
					addTestEvents({ scopeId, startDate: getDate(30), endDate: getDate(90), summary: 'touched end'}),
					addTestEvents({ scopeId, startDate: getDate(-30), endDate: getDate(90), summary: 'start before and end after'}),
					addTestEvents({ scopeId, startDate: getDate(-60), endDate: getDate(-30), summary: 'end before - should not found'}),
					addTestEvents({ scopeId, startDate: getDate(90), endDate: getDate(120), summary: 'start after - should not found'}),
					addTestEvents({ scopeId: scopeIdThatIsNotRequested, startDate: getDate(15), endDate: getDate(45), summary: 'other scope - should not found'}),
					// can not create at the moment -> fix over mock
					// addTestEvents({ scopeId: scopeIdWithoutReadPermissions, startDate: getDate(15), endDate: getDate(45), summary: 'no permissions - should not found'}),
					addTestEvents({ scopeId, startDate: getDate(-7200), endDate: getDate(-7140), summary: 'weekly every monday', frequency: 'WEEKLY', repeat_until: getDate(3600), weekday: ['MO']}),
					addTestEvents({ scopeId, startDate: getDate(-7200), endDate: getDate(-7140), summary: 'weekly every monday - should not found', frequency: 'WEEKLY', repeat_until: getDate(-3600), weekday: ['MO']}),
				]);
			});

			after(async () => {
				resolvedServerScopes.data = resolvedServerScopes.data.filter(scope => scope.id !== scopeId);
				await resetDB();
			});

			it('find all events that touched by requested time box and scope', async () => {
				const result = await request(app)
					.get('/events')
					.query({
						from: getDate(0),
						until: getDate(60),
						'scope-id': scopeId
					})
					.set('Authorization', userId)

					// shoud.include
				expect(result.body.data.some((e) => e.attributes.summary === 'touched start'), 'touched start').to.be.true;
				expect(result.body.data.some((e) => e.attributes.summary === 'in time'), 'in time').to.be.true;
				expect(result.body.data.some((e) => e.attributes.summary === 'touched end'), 'touched end').to.be.true;
				expect(result.body.data.some((e) => e.attributes.summary === 'start before and end after'), 'start before and end after').to.be.true;
				expect(result.body.data.some((e) => e.attributes.summary === 'touched start'), 'touched start').to.be.true;
				expect(result.body.data.some((e) => e.attributes.summary === 'weekly every monday'), 'weekly every monday').to.be.true;

				expect(result.body.data.some((e) => e.attributes.summary === 'end before - should not found'), 'end before - should not found').to.be.false;
				expect(result.body.data.some((e) => e.attributes.summary === 'start after - should not found'), 'start after - should not found').to.be.false;
				expect(result.body.data.some((e) => e.attributes.summary === 'other scope - should not found'), 'other scope - should not found').to.be.false;
				expect(result.body.data.some((e) => e.attributes.summary === 'weekly every monday - should not found'), 'weekly every monday - should not found').to.be.false;
			});
		});

		describe('DELETE by scope', async () => {
			const scopeId = '59cce16281297026d02cde123';
			const anotherScopeId = 'another_id';
			const courseName = 'post test';
			let events;
			before(async () => {
				await addCourseScope(resolvedServerScopes, scopeId, courseName, true);
				events = await Promise.all([
					addTestEvents({ scopeId, startDate: getDate(-30), endDate: getDate(30), summary: 'touched start'}),
					addTestEvents({ scopeId, startDate: getDate(15), endDate: getDate(45), summary: 'in time'}),
					addTestEvents({ scopeId, startDate: getDate(30), endDate: getDate(90), summary: 'touched end'}),
					addTestEvents({ scopeId, startDate: getDate(-30), endDate: getDate(90), summary: 'start before and end after'}),
					addTestEvents({ scopeId, startDate: getDate(-60), endDate: getDate(-30), summary: 'end before - should not found'}),
					addTestEvents({ scopeId, startDate: getDate(90), endDate: getDate(120), summary: 'start after - should not found'}),

					addTestEvents({ scopeId: anotherScopeId, startDate: getDate(90), endDate: getDate(120), summary: 'start after - should not found'}),
					addTestEvents({ scopeId: anotherScopeId, startDate: getDate(90), endDate: getDate(120), summary: 'start after - should not found'}),
					addTestEvents({ scopeId: anotherScopeId, startDate: getDate(90), endDate: getDate(120), summary: 'start after - should not found'}),
					addTestEvents({ scopeId: anotherScopeId, startDate: getDate(90), endDate: getDate(120), summary: 'start after - should not found'})
				]);

			});

			after(async () => {
				resolvedServerScopes.data = resolvedServerScopes.data.filter(scope => scope.id !== scopeId);
				await resetDB();
			});

			it('should successfully delete all events for the scope and not delete other events', async () => {
				const props = [scopeId];
				let resultForScopeToBeDeleted = await db.query('SELECT id from events where scope_id = $1', props);
				let expectedResultNotToBeDeleted = await db.query('SELECT id from events where scope_id != $1', props);
				expect(resultForScopeToBeDeleted.length).to.be.greaterThan(0);
				expect(expectedResultNotToBeDeleted.length).to.be.greaterThan(0);

				const result = await request(app)
					.delete(`/scopes/${scopeId}`)
					.set('Authorization', userId);

				resultForScopeToBeDeleted = await db.query('SELECT id from events where scope_id = $1', props);
				let actualResultNotToBeDeleted = await db.query('SELECT id from events where scope_id != $1', props);

				expect(result.statusCode).to.be.equal(204);
				expect(resultForScopeToBeDeleted.length).to.be.equal(0);
				// check that nothing except expected values was deleted
				expect(actualResultNotToBeDeleted.length).to.be.greaterThan(0);
				expect(expectedResultNotToBeDeleted).to.be.deep.equal(actualResultNotToBeDeleted);
			});
		});
		describe('DELETE Duplicates', async () => {
			const scopeId = '59cce16281297026d02cde123';
			const anotherScopeId = 'another_id';
			const courseName = 'post test';
			let events;
			before(async () => {
				await addCourseScope(resolvedServerScopes, scopeId, courseName, true);
				await addCourseScope(resolvedServerScopes, anotherScopeId, courseName, true);

				events = await Promise.all([
					// last events
					addCourseTestEvent({ scopeId, summary: "orig", weekday: "MO", last_modified: getDate(-30)}),
					addCourseTestEvent({ scopeId, summary: "orig", weekday: "TU", last_modified: getDate(-30)}),
					addCourseTestEvent({ scopeId, summary: "orig", weekday: "WE", last_modified: getDate(-30)}),

					// duplicates
					addCourseTestEvent({ scopeId, summary: "dup", weekday: "MO", last_modified: getDate(-60)}),
					addCourseTestEvent({ scopeId, summary: "dup", weekday: "TU", last_modified: getDate(-60)}),
					addCourseTestEvent({ scopeId, summary: "dup", weekday: "WE", last_modified: getDate(-60)}),

					// other
					addCourseTestEvent({ scopeId: anotherScopeId, summary: "other", weekday: "MO", last_modified: getDate(-30)}),
					addCourseTestEvent({ scopeId: anotherScopeId, summary: "other", weekday: "TU", last_modified: getDate(-30)}),
					addCourseTestEvent({ scopeId: anotherScopeId, summary: "other", weekday: "WE", last_modified: getDate(-30)}),
				]);

			});

			after(async () => {
				resolvedServerScopes.data = resolvedServerScopes.data.filter(scope => scope.id !== scopeId);
				await resetDB();
			});

			it('should successfully delete duplicates and not delete not duplicates', async () => {
				const props = [scopeId, 'dup'];
				let resultsToBeDeleted = await db.query("SELECT id from events where scope_id = $1 and summary = $2", props);
				let expectedResultNotToBeDeleted = await db.query("SELECT id from events where scope_id != $1 or summary != $2", props);
				expect(resultsToBeDeleted.length).to.be.greaterThan(0);
				expect(expectedResultNotToBeDeleted.length).to.be.greaterThan(0);

				const result = await request(app)
					.delete(`/events/duplicates`);

				resultsToBeDeleted = await db.query("SELECT id from events where scope_id = $1 and summary = $2", props);
				let actualResultNotToBeDeleted = await db.query("SELECT id from events where scope_id != $1 or summary != $2", props);

				expect(result.statusCode).to.be.equal(204);
				expect(resultsToBeDeleted.length).to.be.equal(0);
				// check that nothing except expected values was deleted
				expect(actualResultNotToBeDeleted.length).to.be.greaterThan(0);
				expect(expectedResultNotToBeDeleted).to.be.deep.equal(actualResultNotToBeDeleted);
			});
		});

		describe('PUT', () => {
	
		});
	});

	describe('isc', () => {
		describe('CREATE', () => {

		});
	});
});


src/config.js
/**
 * Global Settings fot the Schul-Cloud Calendar Service
 */

const NODE_ENV = process.env.NODE_ENV || 'development';
exports.NODE_ENV = NODE_ENV;
exports.REQUEST_LOGGING_ENABLED = process.env.REQUEST_LOGGING_ENABLED || NODE_ENV === 'development';

/**
 * Database settings for production environment
 */
exports.DB_HOST = process.env.DB_HOST;
exports.DB_DATABASE = process.env.DB_DATABASE;
exports.DB_USERNAME = process.env.DB_USERNAME;
exports.DB_PASSWORD = process.env.DB_PASSWORD;
exports.DB_PORT = process.env.DB_PORT || 5432;
exports.CONNECTION_POOL_SIZE = process.env.CONNECTION_POOL_SIZE || 10;
exports.DB_IDLE_TIMEOUT_MILLIS = process.env.DB_IDLE_TIMEOUT_MILLIS;
exports.DB_MAX_USES = process.env.DB_MAX_USES;

/**
 * Base Paths
 */
const SCHULCLOUD_BASE_PATH = process.env.SCHULCLOUD_BASE_PATH || 'http://localhost:3030';
const NOTIFICATION_SERVICE_BASE_PATH = process.env.NOTIFICATION_SERVICE_BASE_PATH || 'https://schul-cloud.org:3030';
exports.SCHULCLOUD_BASE_PATH = SCHULCLOUD_BASE_PATH;
exports.DOMAIN_NAME = process.env.DOMAIN_NAME || 'schul-cloud.org';
exports.ROOT_URL = process.env.ROOT_URL || `https://${exports.DOMAIN_NAME}:3000`;

/**
* API Paths
*/
exports.API_KEY = process.env.API_KEY || 'example';
const SERVER_USERS_URI = '/resolve/users/';
const SERVER_SCOPES_URI = '/resolve/scopes/';
exports.SERVER_USERS_URI = SERVER_USERS_URI;
exports.SERVER_SCOPES_URI = SERVER_SCOPES_URI;
exports.SCHULCLOUD_ALL_USERS_FOR_UUID = process.env.SCHULCLOUD_ALL_USERS_FOR_UUID || SCHULCLOUD_BASE_PATH + SERVER_USERS_URI;
exports.SCHULCLOUD_ALL_SCOPES_FOR_TOKEN = process.env.SCHULCLOUD_ALL_SCOPES_FOR_TOKEN || SCHULCLOUD_BASE_PATH + SERVER_SCOPES_URI;
exports.NOTIFICATION_SERVICE_NEW_NOTIFICATION = process.env.NOTIFICATION_SERVICE_NEW_NOTIFICATION || NOTIFICATION_SERVICE_BASE_PATH + '/messages';

/**
 * Notification Settings
 */
exports.NOTIFICATION_SERVICE_TOKEN = process.env.NOTIFICATION_SERVICE_TOKEN || 'service1';
exports.NOTIFICATION_SCHULCLOUD_ID = process.env.NOTIFICATION_SCHULCLOUD_ID || '1';

/**
 * CORS
 */
exports.CORS_ORIGIN = process.env.CORS_ORIGIN || 'https://schulcloud.github.io';

/**
 * Logging
 */
exports.LOG_LEVEL = process.env.LOG_LEVEL || 'debug';

/**
 * Setttings
 */
exports.DAY_IN_MS = 1000 * 60 * 60 * 24;
exports.SCOPE_DISPLAY_OLD_EVENTS_FROM_LAST_DAYS = 21;
exports.SCOPE_DISPLAY_OLD_EVENTS_UNTIL_DAYS = 365 * 2;

/**
 * Feature toggles
 */
exports.IS_MIGRATION = process.env.IS_MIGRATION === 'true' || NODE_ENV === 'test';


src/infrastructure/database.js
const pg = require('pg');
const types = require('pg').types;
const INTERVAL_OID = 1186;
const config = require('../config');
const logger = require('../infrastructure/logger');

/*
pg-promise solve the issue with version 10.5.0
https://github.com/vitaly-t/pg-promise/issues/680
{
   message: 'Client has encountered a connection error and is not queryable',
   level: 'error',
   logType: 'error'
}
*/

logger.error('Client has encountered a connection error and is not queryable');

const PostGresErrorLog = (err) => {
	logger.error('[Database]', err);
}

const PostGresInfoLog = (info) => {
	logger.alert('[Database] '+info);
}

const UnexpectedConnectionErrorHandler = (err) => {
	PostGresErrorLog(err);
	process.exit(config.PROCESS_EXIT.DATABASE_FAIL);
}

const restart = (client, connect) => {
	try {
		client.end().then(() => {
			client = connect(db).catch(UnexpectedConnectionErrorHandler);
		})
	} catch (err) {
		UnexpectedConnectionErrorHandler(err);
	}
}


let client = null;
const connect = (db) => {
	const c = new pg.Client(db);

	return c.connect().then(() => {
		PostGresInfoLog('Postgres DB connected.');
		client = c;

		client.on('error', (err) => {
			PostGresErrorLog(err);
			if (typeof err === 'object' &&  err.severity === 'FATAL') {
					restart(client, connect);
			}
		});

		client.query("SET intervalstyle = 'iso_8601';");
		return client;
	}).catch(UnexpectedConnectionErrorHandler);
}

const resolveDBCredentials = () => {
	let db;
	if (config.NODE_ENV === 'production' || config.NODE_ENV === 'test') {
		const db_username = config.DB_USERNAME;
		const db_password = config.DB_PASSWORD;
		const db_host = config.DB_HOST;
		const db_port = config.DB_PORT;
		const db_database = config.DB_DATABASE;
	
		if (!db_username ||
			(config.NODE_ENV === 'production' && !db_password) ||
			!db_host ||
			!db_port ||
			!db_database) {
			throw new Error('expected database credentials but incomplete specified');
		}
	
		if (db_password) {
			db = {
				user: db_username,
				password: db_password,
				host: db_host,
				port: db_port,
				database: db_database,
			};
		} else {
			db = {
				user: db_username,
				host: db_host,
				port: db_port,
				database: db_database,
			};
		}
	} else {
		db = {
			user: 'node',
			password: 'node',
			host: 'localhost',
			port: 5432,
			database: 'schulcloud_calendar',
		};
	}
	return db;
}


const prom = connect(resolveDBCredentials());

// TODO link for explain this action
types.setTypeParser(INTERVAL_OID, function(val) {
	return val.toString();
});

// return a get client function getClient()
module.exports = (promise = false) => promise ? prom : client;


src/security/authentication.js
const getAllScopesForToken = require('../http/getAllScopesForToken');
const config = require('./../config');

function authenticateFromHeaderField(req, res, next) {
	const token = req.get('Authorization');
	getUser(req, res, next, token);
}

function authenticateFromQueryParameter(req, res, next) {
	const token = req.query['authorization'];
	getUser(req, res, next, token);
}

function isMigration(req, res, next) {
	if (config.IS_MIGRATION === true) {
		next();
	} else {
		const err = {
			message: 'This route is available only for migration!',
			status: 401,
			title: 'Unauthorized',
		};
		next(err);
	}
}

function getUser(req, res, next, token) {
	if (token) {
		req.token = token;
		getAllScopesForToken(token)
			.then((apiResponse) => {
				req.user = parseUserInformation(apiResponse);
				next();
			})
			.catch(() => {
				const err = {
					message: 'Invalid Authorization token!',
					status: 401,
					title: 'Unauthorized',
				};
				next(err);
			});
	} else {
		const err = {
			message: 'Missing Authorization token!',
			status: 401,
			title: 'Unauthorized',
		};
		next(err);
	}
}

function parseUserInformation(apiResponse) {
	apiResponse = JSON.parse(apiResponse);

	let user = {};
	user.scopes = {};
	apiResponse.data.forEach(function (scope) {
		if (scope.type === 'user') {
			user.id = scope.id;
			user.name = scope.attributes.name;
		}

		// These scopes also include the user ID
		user.scopes[scope.id] = {};
		user.scopes[scope.id].id = scope.id;
		user.scopes[scope.id].name = scope.attributes.name;
		user.scopes[scope.id].authorities = {};
		scope.attributes.authorities.forEach(function (authority) {
			user.scopes[scope.id].authorities[authority] = true;
		});
	});
	return user;
}

module.exports = {
	authenticateFromHeaderField,
	authenticateFromQueryParameter,
	isMigration,
};


src/queries/events/deleteEvents.js
const db = require('../../infrastructure/databasePromise');
const { allColumns } = require('./constants');

async function deleteEvents(eventId, scopeIds) {
	const query = `DELETE FROM events WHERE event_id = $1 AND scope_id = ANY($2) RETURNING ${allColumns}`;
	const params = [eventId, scopeIds];

	return db.query(query, params);
}

async function deleteDuplicatesForCourses() {
	const query = `DELETE FROM events where id in (
		SELECT e.id from
		(
		-- events grouped by scope_id
		select scope_id, max("last-modified") as maxLastMod
		FROM events 
		WHERE x_fields ->> 'x-sc-courseId' is not null and repeat_wkst is not null
		GROUP BY scope_id) as t1
		-- return only events which were modified before max last modified date - 20 seconds
		JOIN events as e
			 ON t1.scope_id = e.scope_id and e."last-modified" < (t1.maxLastMod - '20 seconds'::interval)
	) RETURNING ${allColumns}`;
	return db.query(query);
}

async function deleteEventsForScope(scopeId) {
	const query = `DELETE FROM events WHERE scope_id = $1 RETURNING ${allColumns}`;
	const params = [scopeId];

	return db.query(query, params);
}

module.exports = {
	deleteEvents,
	deleteEventsForScope,
	deleteDuplicatesForCourses,
};


src/services/events/deleteDuplicatedEvents.js
const { deleteDuplicatesForCourses } = require('../../queries/events/deleteEvents');
const logger = require('../../infrastructure/logger');

duplicatesToString = (result) => {
    return result.map((d) => `${d.scope_id} => ${d.event_id}`).join(';');
}

deleteDuplicatedEvents = async () => {
    try {
        logger.alert(`Starting deleting duplicated events for courses`);
        const result = await deleteDuplicatesForCourses();
        const count = result.length;
        logger.alert(`Successfully deleted ${count} duplicated events for courses`);
        logger.alert(`Deleted duplicates: ${duplicatesToString(result)}`);
        return { count, data: result };
    } catch (e) {
        logger.alert(`Deleting of duplicated events failed: ${JSON.stringify(e)}`, e);
        throw e;
    }
}

module.exports = {
    deleteDuplicatedEvents
};


src/routes/events.js
const express = require('express');
const bodyParser = require('body-parser');
const router = express.Router();
const cors = require('cors');
const config = require('../config');
const corsOptions = {
	origin: config.CORS_ORIGIN
};
router.use(cors(corsOptions));
router.use(bodyParser.json());
router.use(bodyParser.urlencoded({ extended: false }));

// authentication, authorization and preprocessing
const { authenticateFromHeaderField, isMigration } = require('../security/authentication');
const { authorizeAccessToScopeId, authorizeAccessToObjects, authorizeWithPotentialScopeIds } = require('../security/authorization');
const jsonApiToJson = require('../parsers/event/jsonApiToJson');
const icsToJson = require('../parsers/event/icsToJson');

// response
const returnSuccess = require('../utils/response/returnSuccess');
const sendNotification = require('../services/sendNotification');
const eventsToJsonApi = require('../parsers/event/eventsToJsonApi');
const eventsToIcsInJsonApi = require('../parsers/event/eventsToIcsInJsonApi');

// content
const getEvents = require('../services/events/getEvents');
const getOriginalEvent = require('../queries/original-events/getOriginalEvent');
const insertEvents = require('../services/events/insertEvents');
const { deleteDuplicatedEvents } = require('../services/events/deleteDuplicatedEvents');
const { deleteEventWithScope, deleteAllEventsForScope } = require('../services/events/deleteEventWithScope');
const updateEvents = require('../services/events/updateEvents');

/* routes */

router.get('/events', authenticateFromHeaderField, function (req, res, next) {
	const filter = {
		scopeId: req.query['scope-id'],
		eventId: req.query['event-id'],
		from: req.query['from'],
		until: req.query['until'],
		all: req.query['all']
	};

	if (['1', 1, 'true'].includes(filter.all)) {
		filter.all = true;
	} else if(['0', 0, 'false'].includes(filter.all)) {
		filter.all = false;
	}

	const user = req.user;
	authorizeAccessToScopeId(user, filter.scopeId)
		.then(() => getEvents(filter, user.scopes))
		.then((events) => authorizeAccessToObjects(user, 'can-read', events))
		.then(eventsToJsonApi)
		.then((jsonApi) => { returnSuccess(res, 200, jsonApi); })
		.catch(next);
});

router.post('/events', jsonApiToJson, authenticateFromHeaderField, function (req, res, next) {
	handlePost(req, res, next, eventsToJsonApi);
});

router.post('/events/ics', icsToJson, authenticateFromHeaderField, function (req, res, next) {
	handlePost(req, res, next, eventsToIcsInJsonApi);
});

function handlePost(req, res, next, outputFormatter) {
	const user = req.user;
	const events = req.events;

	authorizeAccessToObjects(user, 'can-write', events)
		.then((events) => doInserts(events, user))
		.then(sendInsertNotification)
		.then(outputFormatter)
		.then((jsonApi) => { returnSuccess(res, 200, jsonApi); })
		.catch(next);
}

function doInserts(events, user) {
	return new Promise(function (resolve, reject) {
		insertEvents(events, user)
			.then(resolve)
			.catch(reject);
	});
}

function sendInsertNotification(insertedEvents) {
	insertedEvents.forEach((insertedEvent) => {
		const {scope_id, summary, dtstart, dtend} = insertedEvent;
		sendNotification.forNewEvent(scope_id, summary, dtstart, dtend);
	});
	return insertedEvents;
}

router.put('/events/:eventId', jsonApiToJson, authenticateFromHeaderField, function (req, res, next) {
	handlePut(req, res, next, eventsToJsonApi);
});

router.put('/events/ics/:eventId', icsToJson, authenticateFromHeaderField, function(req, res, next) {
	handlePut(req, res, next, eventsToIcsInJsonApi);
});

function handlePut(req, res, next, outputFormatter) {
	const eventId = req.params.eventId;
	const event = req.events[0];
	const scopeIds = event.scope_ids;
	const user = req.user;

	authorizeWithPotentialScopeIds(eventId, scopeIds, user, getEvents, getOriginalEvent, 'eventId')
		.then(() => doUpdates(event, eventId))
		.then(sendUpdateNotification)
		.then(outputFormatter)
		.then((jsonApi) => { returnSuccess(res, 200, jsonApi); })
		.catch(next);
}

function doUpdates(event, eventId) {
	return new Promise(function (resolve, reject) {
		updateEvents(event, eventId)
			.then((updatedEvents) => {
				if (updatedEvents.length === 0) {
					const error = new Error('Given eventId or scopeIds not found '
						+ 'for event modification');
					error.status = 404;
					error.title = 'Query Error';
					reject(error);
				} else {
					resolve(updatedEvents);
				}
			})
			.catch(reject);
	});
}

function sendUpdateNotification(updatedEvents) {
	updatedEvents.forEach((updatedEvent) => {
		const { scope_id, summary, dtstart, dtend } = updatedEvent;
		sendNotification.forModifiedEvent(scope_id, summary, dtstart, dtend);
	});
	return updatedEvents;
}

router.delete('/events/duplicates', isMigration, (req, res, next) => {
	deleteDuplicatedEvents()
		.then((result) => ({data: result}))
		.then((jsonApi) => { returnSuccess(res, 204, jsonApi); })
		.catch(next);
});
router.delete('/scopes/:scopeId', authenticateFromHeaderField, (req, res, next) => {
	const scopeId = req.params.scopeId;
	const user = req.user;

	authorizeAccessToScopeId(user, scopeId)
		.then(() => deleteAllEventsForScope(scopeId))
		.then((result) => ({data: result}))
		.then((jsonApi) => { returnSuccess(res, 204, jsonApi); })
		.catch(next);
});

const processDeletedEvents = (deletedEvents, res, next) => {
	if (deletedEvents.length > 0) {
		returnSuccess({data: deletedEvents}, 204);
	} else {
		const err = {
			message: 'Given eventId or scopeIds not found for event deletion',
			status: 404,
			title: 'Query Error',
		};
		next(err);
	}
}

router.delete('/events/:eventId', jsonApiToJson, authenticateFromHeaderField, function (req, res, next) {
	const eventId = req.params.eventId;
	const user = req.user;
	const scopeIds = user.scopes;

	authorizeWithPotentialScopeIds(eventId, scopeIds, user, getEvents, getOriginalEvent, 'eventId')
		.then(() => deleteEventWithScope(eventId, scopeIds))
		.then((deletedEvents) => {
			processDeletedEvents(deletedEvents, res, next);
		})
		.catch(next);
});

module.exports = router;
